### PR TITLE
Password reset

### PR DESCRIPTION
**Pull request by kalpeshDocFox**
Hi Gents , Please assist with a review for CH12

### CODE REVIEW

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by(email: params[:email])
+    if user && !user.activated? && user.authenticated?(:activation, params[:id])
```
Comment (c13e) [inline app/controllers/account_activations_controller.rb]: try extracting such chained conditionals into well named private method, ex: `can_activate_account?`

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: [:edit, :update] 
+  before_action :valid_user, only: [:edit, :update] 
+  before_action :check_expiration, only: [:edit, :update] # Case (1)
+ 
+  def new 
+  end
```
Comment (c13e) [inline app/controllers/password_resets_controller.rb]: nuke action if not used, please also add a line between methods

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
```
Comment (c13e) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    @user = User.find_by(email: email_param) 
```
always best to optimize for readability by extracting non-essential details into methods/variables

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    @user = User.find_by(email: params[:password_reset][:email].downcase) 
+    if @user
+    @user.create_reset_digest
+    @user.send_password_reset_email
+    flash[:info] = "Email sent with password reset instructions" 
+    redirect_to root_url
+    else
+    flash.now[:danger] = "Email address not found"
+    render 'new'
+    end
```
Comment (c13e) [inline app/controllers/password_resets_controller.rb]:
```suggestion
    if @user
      @user.create_reset_digest
      @user.send_password_reset_email
      flash[:info] = "Email sent with password reset instructions" 
      redirect_to root_url
    else
      flash.now[:danger] = "Email address not found"
      render 'new'
    end
```

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  def edit
+  end
```
Comment (c13e) [inline app/controllers/password_resets_controller.rb]: ditto, nuke if unused

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      end
+    end
+
+    # Checks expiration of reset token.
+    def check_expiration
+      if @user.password_reset_expired?
+      flash[:danger] = "Password reset has expired."
+      redirect_to new_password_reset_url
+      end 
+    end
```
Comment (c13e) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  private

  def user_params
    params.require(:user).permit(:password, :password_confirmation)
  end

  # Before filters

  def get_user
  @user = User.find_by(email: params[:email])
  end
  # Confirms a valid user.
  def valid_user
    unless (@user && @user.activated? && @user.authenticated?(:reset, params[:id])) 
    redirect_to root_url
    end
  end

  # Checks expiration of reset token.
  def check_expiration
    if @user.password_reset_expired?
    flash[:danger] = "Password reset has expired."
    redirect_to new_password_reset_url
    end 
  end
```
indentation is off here

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -10,25 +10,45 @@ def can_authenticate?(user)
...
-      forwarding_url = session[:forwarding_url]
-      reset_session
-      log_in @user
-      params[:session][:remember_me] == '1' ? remember(@user) : forget(@user)
-      session[:session_token] = @user.session_token
-      redirect_to forwarding_url || @user
-    else
-      flash.now[:danger] = 'Invalid email/password combination'
-      render 'new'
+      if @user.activated?
```
Comment (c13e) [inline app/controllers/sessions_controller.rb:27]: can you extract the logic in the if & else branches into methods to improve readability since we have nested conditionals here?

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -10,25 +10,45 @@ def can_authenticate?(user)
...
     end
   end
-
+  # Returns the current logged-in user (if any).
   def current_user
-    if session[:user_id]
-      User.find_by(id: session[:user_id])
+    if (user_id = session[:user_id])
+      user ||= User.find_by(id: user_id)
+        @current_user||= user if (user.session_token!=session['session_token'])
```
Comment (c13e) [inline app/controllers/sessions_controller.rb]:
```suggestion
      @current_user ||= user if (user.session_token != session['session_token'])
```
I would also try to extract everything after the `if` into a method.

Although reading the implementation here, I'm a bit confused, shouldn't we set `@current_user` if the session token matches (instead of current implementation where it's set when it mismatches)?

File: app/mailers/user_mailer.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+class UserMailer < ApplicationMailer
+
```
Comment (c13e) [inline app/mailers/user_mailer.rb]: nuke extra line here

File: app/models/user.rb [hunk 10]
```diff
@@ -44,19 +53,52 @@ def self.new_token
...
+    UserMailer.account_activation(self).deliver_now
+  end
+
+  # Sets the password reset attributes.
+  def create_reset_digest
+    self.reset_token = User.new_token 
+    update_columns(reset_digest: User.digest(reset_token), reset_sent_at:Time.zone.now)
+  end
+ 
+
```
Comment (c13e) [inline app/models/user.rb]: nuke extra line

File: app/views/password_resets/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        <%= f.label :password %>
+        <%= f.password_field :password, class: 'form-control' %>
+
+        <%= f.label :password_confirmation, "Confirmation" %>
+        <%= f.password_field :password_confirmation, class: 'form-control' %>
+        
+        <%= f.submit "Update password", class: "btn btn-primary" %> 
+        <% end %>
+    </div>
+</div>
```
Comment (c13e) [inline app/views/password_resets/edit.html.erb:19]: we-use 2-space indentation for template files too :)

File: test/controllers/account_activations_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class AccountActivationsControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
Comment (c13e) [inline test/controllers/account_activations_controller_test.rb]: are the tests not implemented in this chapter?

File: test/integration/password_resets_test.rb [hunk 13]
```diff
@@ -0,0 +1,69 @@
+require 'test_helper'
+class PasswordResetsTest < ActionDispatch::IntegrationTest
+  def setup 
+    ActionMailer::Base.deliveries.clear 
+    @user = users(:michael)
+  end
+
+  test "password resets" do
```
Comment (c13e) [inline test/integration/password_resets_test.rb]: there's a few issues with this test case:
  * non-descriptive test name
  * it's doing a lot, hard to follow & hard to debug when something breaks

please try to break each scenario (ex: each request + assertions) into its own test case with a descriptive title

File: test/integration/user_show_test.rb [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+        assert_response :redirect
+        assert_redirected_to root_url
+    end
+    
+    test "should display user when activated" do
+        get user_path(@activated_user)
+        assert_response :success
+        assert_template "users/show"
+    end 
+end
```
Comment (c13e) [inline test/integration/user_show_test.rb:19]: please fix indentation & try to set it to 2-spaces within your editor config

File: test/integration/users_index_test.rb [hunk 15]
```diff
@@ -11,22 +11,25 @@ def setup
...
-    first_page_of_users.each do |user|
-      assert_select 'a[href=?]', user_path(user), text: user.first_name
-      unless user == @admin
-        assert_select 'a[href=?]', user_path(user), text: 'delete'
-      end
-    end
-    assert_difference 'User.count', -1 do
-      delete user_path(@non_admin)
+    assigns(:users).each do |user|
+        assert user.activated
```
Comment (c13e) [inline test/integration/users_index_test.rb]: ditto indentation

File: test/integration/users_login_test.rb [hunk 16]
```diff
@@ -25,8 +25,11 @@ def setup
 
   test "login with valid information followed by logout" do
     get login_path
-    post login_path, params: { session: { email: @user.email,
-                                          password: 'password' } }
+    post login_path, params: { session: 
+                              { email: @user.email,
+                                password: 'password' 
+                                } 
+                              }
```
Comment (c13e) [inline test/integration/users_login_test.rb]: ditto extrating params into variable (see comment on previous PR and suggestion in next comment below)

File: test/integration/users_signup_test.rb [hunk 17]
```diff
@@ -22,8 +27,36 @@ class UsersSignupTest < ActionDispatch::IntegrationTest
         confirm_password: "password" } }
     end
     follow_redirect!
-    assert_template 'users/show'
-    assert is_logged_in?
+    #assert_template 'users/show'
+    #assert is_logged_in?
```
Comment (c13e) [inline test/integration/users_signup_test.rb:31]: why comment these out?

File: test/integration/users_signup_test.rb [hunk 18]
```diff
@@ -22,8 +27,36 @@ class UsersSignupTest < ActionDispatch::IntegrationTest
...
+  test "valid signup information with account activation" do get signup_path
+    assert_difference 'User.count', 1 do
+    post users_path, params: { user: 
+                                { name: "Example User",
+                                  email: "user@example.com",
+                                  password:              "password",
+                                  password_confirmation: "password" 
+                                } 
+                              }
+    end
```
Comment (c13e) [inline test/integration/users_signup_test.rb]:
```suggestion
    assert_difference 'User.count', 1 do
      params = {
        user: {
          name: "Example User",
          email: "user@example.com",
          password: "password",
          password_confirmation: "password"
        }
      }
      post users_path, params
    end
```

File: test/integration/users_signup_test.rb [hunk 19]
```diff
@@ -22,8 +27,36 @@ class UsersSignupTest < ActionDispatch::IntegrationTest
...
+    # Valid token, wrong email
+    get edit_account_activation_path(user.activation_token, email: 'wrong') 
+    assert_not is_logged_in?
+    # Valid activation token
+    get edit_account_activation_path(user.activation_token, email: user.email) 
+    assert user.reload.activated?
+    follow_redirect!
+    assert is_logged_in?
+    end 
+
```
Comment (c13e) [inline test/integration/users_signup_test.rb]: nuke extra line

File: test/mailers/previews/user_mailer_preview.rb [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    user.activation_token = User.new_token 
+    UserMailer.account_activation(user)
+  end
+
+  # Preview this email at
+  # http://localhost:3000/rails/mailers/user_mailer/password_reset 
+  def password_reset
+      user = User.first
+      user.reset_token = User.new_token
+      UserMailer.password_reset(user)
```
Comment (c13e) [inline test/mailers/previews/user_mailer_preview.rb]: indentation

File: test/integration/users_signup_test.rb (hunk 17 above)
Comment (kalpeshDocFox) [reply test/integration/users_signup_test.rb:31]: In CH11  on page 647 it was commented out, reason was the create method within the controller was modified to redirect to root url and flat a email message instead of logging in and redirecting to logged in user
